src/TorchSharp/DataLoader.cs

```diff
@@ -111,17 +121,50 @@ public bool MoveNext()
                         {
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
-                            List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
+
+                            var tensorIndexList = new List<long> {currentVal};
+                            for (int i = 1; i < batchSize; i++) {
                                 if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
+                                tensorIndexList.Add(currentVal);
                             }
 
+                            if (!shuffle)
+                                tensorIndexList.Reverse(); // ConcurrentBag::Add inserts data into first of the collections. So, I'll reverse them.
+
+                            ConcurrentBag<Dictionary<string, Tensor>> dic = new();
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
+
+                            CreateBatch(null);
+
+                            while (dic.Count < tensorIndexList.Count) { } //Wait for tensors ready
+
                             Current = new();
-                            foreach (var x in dic[0].Keys)
+                            foreach (var x in dic.First().Keys)
                                 Current[x] = cat(dic.Select(k => k[x].unsqueeze(0)).ToArray(), 0).to(device);
                             return true;
+
+                            void CreateBatch(object _)
+                            {
+                                while (true) {
+                                    var idx = TaskBatch();
+                                    if (idx is null) break;
+                                    dic.Add(dataset.GetTensor(idx.Value));
+                                }
+                            }
+
+                            long? TaskBatch()
+                            {
+                                lock (taskBatchLock) {
+                                    if (taskedBatchCount < tensorIndexList.Count)
+                                        return tensorIndexList[taskedBatchCount++];
+                                    return null;
+                                }
```

Comment:
You don't need a lock here. `Interlocked.Increment` should do.

Reply:
In changed code, I think lock is required because taskedBatchCount used with `(taskedBatchCount, tensorIndexList[taskedBatchCount++])`. Interlocked.Increment will not right way.

Reply:
The following code should be perfectly safe and avoid having the lock:

```C#
var tbc = System.Threading.Interlocked.Increment(ref taskedBatchCount);
if (taskedBatchCount <= tensorIndexList.Count)
    return (taskedBatchCount-1, tensorIndexList[tbc]);
return null;
```